### PR TITLE
overlong filtering

### PR DESCRIPTION
**Pull request by shirinyamani**
# What does this PR do?

In RL training, we typically set a maximum length for generation, with overlong samples truncated accordingly. [DAPO](https://huggingface.co/papers/2503.14476) paper found that improper reward shaping for truncated samples can introduce reward noise and significantly disrupt the training process. To investigate the impact of this reward noise, [DAPO](https://huggingface.co/papers/2503.14476) paper first applied an Overlong Filtering strategy which masks the loss of truncated samples. They find that this approach significantly stabilizes training and enhances
performance. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by qgallouedec — COMMENTED** [review]
Looks good!!
A few comments/questions 

- ideally we would like some benchmarks to validate the results presented in the DAPO paper. But IMO it's ok to merge before we have these though
- can you add a unittest for this setting?
- what happens if all completions are truncated?
- should we exclude the samples from the reward calculation as well? So that it does bias the reward normalization?
- in the doc could you mention that this principle was added in the DAPO paper?
- could you add the new metrics to the "logged metrics" section of the GRPO doc

**Review by edbeeching — COMMENTED** [review]
Thanks for the implementation, I will launch some benchmarks with it this morning.

I agree it makes sense to exclude truncated sequences from the advantage calculation as well. But perhaps that can be split into another PR?

Another comment, I think the implementation could be simplified by just directly masking the `completion_mask` in the  `_generate_and_score_completions` method, rather than having an additional `truncated_sampled` tensor. This would mean `compute_loss` will be exactly the same and the implementation should be compatible with the liger PR.

File: trl/trainer/grpo_config.py [hunk 1]
```diff
@@ -115,6 +115,10 @@ class GRPOConfig(TrainingArguments):
             applied. The [Dr. GRPO](https://github.com/sail-sg/understand-r1-zero/blob/main/understand-r1-zero.pdf)
             paper recommends not scaling the rewards, as scaling by the standard deviation introduces a question-level
             difficulty bias.
+        mask_truncated_samples (`bool`, *optional*, defaults to `False`):
```
Comment (lewtun) [inline trl/trainer/grpo_config.py:118]: turbo nit: since we use "completions" elsewhere in the config, let's align it here too
```suggestion
        mask_truncated_completions (`bool`, *optional*, defaults to `False`):
```

File: trl/trainer/grpo_config.py [hunk 2]
```diff
@@ -115,6 +115,10 @@ class GRPOConfig(TrainingArguments):
...
+            Whether to mask the loss for samples that were truncated due to exceeding the maximum completion length.
+            When enabled, truncated completions are excluded from loss calculation, preventing them from being
```
Comment (lewtun) [inline trl/trainer/grpo_config.py:120]:
```suggestion
            When enabled, truncated completions are excluded from the loss calculation, preventing them from being
```

File: trl/trainer/grpo_config.py [hunk 3]
```diff
@@ -297,6 +301,14 @@ class GRPOConfig(TrainingArguments):
             "deviation introduces a question-level difficulty bias."
         },
     )
+    mask_truncated_samples: bool = field(
```
Comment (lewtun) [inline trl/trainer/grpo_config.py:304]: Ditto here

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -751,6 +752,15 @@ def _generate_and_score_completions(
         eos_idx[is_eos.any(dim=1)] = is_eos.int().argmax(dim=1)[is_eos.any(dim=1)]
         sequence_indices = torch.arange(is_eos.size(1), device=device).expand(is_eos.size(0), -1)
         completion_mask = (sequence_indices <= eos_idx.unsqueeze(1)).int()
+        # Identify truncated samples (completions that reached max_completion_length without generating an EOS token)
+        truncated_samples = torch.zeros(is_eos.size(0), dtype=torch.bool, device=device)
+        if self.mask_truncated_samples:
+            # A sample is truncated if it has no EOS token and its length equals max_completion_length
+            truncated_samples = ~is_eos.any(dim=1) & (completion_ids.size(1) == self.max_completion_length)
```
Comment (lewtun) [inline trl/trainer/grpo_trainer.py:759]: Is the second requirement necessary? I'm trying to think of a case where a completion would have no EOS token, yet be shorter than `max_completion_length`

File: trl/trainer/grpo_trainer.py [hunk 5]
```diff
@@ -751,6 +752,15 @@ def _generate_and_score_completions(
...
+            # Log the % of truncated samples 
+            truncated_pct = truncated_samples.float().mean().item() * 100
```
Comment (lewtun) [inline trl/trainer/grpo_trainer.py:761]: If I'm not mistaken, we already have this metric as `completions/clipped_ratio` from #3188

File: trl/trainer/grpo_trainer.py (hunk 4 above)
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:759]: I don't think the second requirement is necessary either

File: docs/source/grpo_trainer.md [hunk 6]
```diff
@@ -126,7 +126,8 @@ $$
 \text{clip}\left( \frac{\pi_\theta(o_{i,t} \mid q, o_{i,< t})}{\pi_{\theta_{\text{old}}}(o_{i,t} \mid q, o_{i,< t})}, 1 - \epsilon, 1 + \epsilon \right)
 $$
 A higher value means more tokens were affected by clipping, limiting how much the policy can change.
-
+- `mask_truncated_completions`: The completions that were truncated due to exceeding the maximum completion length. Therefore, excluded from the loss calculation. This feature was introduced as a metric of training stability in the (DAPO paper)[https://huggingface.co/papers/2503.14476]
+- `epsilon_high`: The upper bound of the clipping range. If set to higher value (e.g. 0.28), the policy can have more exploration ability.
```
Comment (qgallouedec) [inline docs/source/grpo_trainer.md:130]: is it logged somewhere?

File: docs/source/grpo_trainer.md [hunk 7]
```diff
@@ -126,7 +126,8 @@ $$
 \text{clip}\left( \frac{\pi_\theta(o_{i,t} \mid q, o_{i,< t})}{\pi_{\theta_{\text{old}}}(o_{i,t} \mid q, o_{i,< t})}, 1 - \epsilon, 1 + \epsilon \right)
 $$
 A higher value means more tokens were affected by clipping, limiting how much the policy can change.
-
+- `mask_truncated_completions`: The completions that were truncated due to exceeding the maximum completion length. Therefore, excluded from the loss calculation. This feature was introduced as a metric of training stability in the (DAPO paper)[https://huggingface.co/papers/2503.14476]
```
Comment (qgallouedec) [inline docs/source/grpo_trainer.md:129]: I think you don't log it anymore
